build/Shared/TaskResultCache.cs

```diff
@@ -0,0 +1,126 @@
+        }
+
+        /// <summary>
+        /// Gets the cached async operation associated with the specified key, or runs the operation asynchronously and returns <see cref="Task{TValue}" /> that the caller can await, and optionally refreshes the cache.
+        /// </summary>
+        /// <param name="key">The key for the async operation to get or store in the cache.</param>
+        /// <param name="refresh"><see langword="true" /> to force the specified asynchronous operation to be executed and stored in the cache even if a cached operation exists, otherwise <see langword="false" />.</param>
+        /// <param name="valueFactory">A <see cref="Func{TResult}" /> to execute asynchronously if a cached operation does not exist.</param>
+        /// <returns>A <see cref="Task{TResult}" /> for the specified asynchronous operation from the cache if found, otherwise the scheduled asynchronous operation to await.</returns>
+        public Task<TValue> GetOrAddAsync(TKey key, bool refresh, Func<Task<TValue>> valueFactory)
+        {
+            return GetOrAddAsync(key, refresh, valueFactory, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Gets the cached async operation associated with the specified key, or runs the operation asynchronously and returns <see cref="Task{TValue}" /> that the caller can await, and optionally refreshes the cache.
+        /// </summary>
+        /// <param name="key">The key for the async operation to get or store in the cache.</param>
+        /// <param name="refresh"><see langword="true" /> to force the specified asynchronous operation to be executed and stored in the cache even if a cached operation exists, otherwise <see langword="false" />.</param>
+        /// <param name="valueFactory">A <see cref="Func{TResult}" /> to execute asynchronously if a cached operation does not exist.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken" /> to use for signaling that an operation should be cancelled.</param>
+        /// <returns>A <see cref="Task{TResult}" /> for the specified asynchronous operation from the cache if found, otherwise the scheduled asynchronous operation to await.</returns>
+        public Task<TValue> GetOrAddAsync(TKey key, bool refresh, Func<Task<TValue>> valueFactory, CancellationToken cancellationToken)
+        {
+            if (!refresh && _cache.TryGetValue(key, out Task<TValue>? result))
+            {
+                return result;
+            }
+
+            lock (_cache)
+            {
+                if (!refresh && _cache.TryGetValue(key, out result))
+                {
+                    return result;
+                }
+
+                result = valueFactory().ContinueWith(TaskComplete, state: null, cancellationToken, TaskContinuationOptions.RunContinuationsAsynchronously, TaskScheduler.Default);
+
+                _cache[key] = result;
```

Comment:
It grows unbounded during a restore yes.  Each restore gets a new context which starts with a new cache.

Reply:
I'm assuming there is no risk since this concept of a dictionary containing cached task results has existed in NuGet for quite some time.  In some of the largest restores where NuGet is restoring thousands of packages, it hasn't been an issue.  The idea here is NuGet is restoring hundreds of projects at the same time and each target framework for each project is walking the dependency graph and each thread wants to resolve a particular package that might involved querying the feed, downloading the `.nupkg`, and unzipping it.  We want each caller to be awaiting the same task for any particular package and not duplicate the work.  Are you thinking of another way to accomplish this that wouldn't grow an unbounded cache?  Would love to hear your thoughts in general of this design.

Reply:
The concern here is that you will continue to hold onto the memory used by these tasks. Given we see this consume 100 GB VM today in some traces, this is a concern.